### PR TITLE
Final Submission

### PR DESCRIPTION
**Pull request by scottdelly**
This submission has the ability to show/hide and toggle the banner. The banner is present on all tabs, and modals appear over the banner.

The dining mode appears when the banner is dragged up or tapped. Dining mode satisfies all specced requirements, but it does not expand dishes in the top dishes section. I am testing out some code to achieve this in a similar way to google inbox and found a promising library on github.

Please let me know if you have any questions or feedback. Thanks!

### CODE REVIEW

**Comment by olarivain** [comment]
Hey, thanks for the PR.
It does solve the problem, so congrats on that, I'll put some more general comments in here.

Some pieces are pretty complicated to follow, for instance: 
```swift
var dishes: [Dish]? {
        get {
            if let dishSection = self.diningModeSections.first(where: { (section) -> Bool in
                switch section { case .dishes: return true default: return false }
            }) {
                if case let .dishes(dishArray) = dishSection {
                    return dishArray
                }
            }
            return nil
        }
    }
```
You have to find the section, then unwrap the data, so you can finally return the array that is actually initially coming from ```self.reservation.restaurant.dishes```. Feels a bit overkill, what prompted you to walk down such a path?

```DiningModeInteractiveTransition``` is a bit hard to follow too. The confusing thing about it to me is that the transition object is the one responsible for triggering the action that will start itself. Feels a bit meta, could you explain the thought process here?

Last but not least, a more ground to earth question: your dishes cell sizing implementation is a close enough approximation (would work great for a prototype), but the last cell is bit chopped on an SE, and quite a bit of room to spare on a 7 Plus. How could address this problem?

**Comment by scottdelly** [comment]
Great questions, I'll address them one-by-one in order:
1) The dishes var is an attempt to expose an array of dishes to the table view that is responsible for displaying them. While the internal logic could be simplified, the idea is that the collection view's array of sections is the single source-of-truth for the entire view. An alternate, and more legible way to write this might be to store the dishes array for the table view separately and simply copy the dishes into it when setting up the collection view sections. This could however lead to some issues where the wrong dishes are displayed for a restaurant if the data is not handled properly.

2) I think you're referring to the lines: 
```
        switch gestureRecognizer.state {
        case .began:
            self.transitionMode.execute()
```
Which confused me at first too. This is according to some existing work I found online, and apparently how UIKit intends for interactive transitions to be started. Essentially it `execute`s the transition (present modally) at the beginning. Since a UIPercentDrivenInteractiveTransition is present (see `interactionControllerForPresentation`), it informs the animation logic that it should not pop-up like a normal modal, but instead wait for calls to the `update` and `finish` or `cancel` methods which happen inside the gesture recognizer. The logic could be polished up a bit for sure, but the idea behind it is purely encapsulation. That is, if we use the standard UIPercentDrivenInteractiveTransition, then the actual dining mode VC can be a simple UIViewController and its parent can simply call present() and all the animation logic is encapsulated elsewhere. 

As far at the `MainTabBarController` is concerned, I think there is some cleanup and simplification that can be done there as well, in terms of the presentation and dismissal logic for the dining mode vc. It's pretty clear that that class went through a few architecture iterations :) 

3) The cell sizing is static at the moment, which works well enough for the first 2 collection view cells but not for the 3rd (table view). There are a few ideas I have about that: One way is to keep it a table view, and have some method that calculates the height of each table view cell, given the content and then adds that up to determine the height of the collection view cell before rendering. Another way is to make each cell and collection view cell, and eliminate the tableview altogether, though this may present some layout and data source weirdness. I think that, starting over, I'd try the second approach as it seems to involve less manual computation. But if this were a real-world scenario I'd try to fix as is, before making architectural changes, and try to get an better understand of the next iteration or two of the design of that whole view before making a decision.

I hope this answers your questions, if not I'm happy to hop on a call to discuss
